File: ch06/py/ll_queue.py
"""
A queue implementation backed by a linked list with a tail pointer. An adapter pattern!
"""

from ch06.py.linked_list import LinkedList


class Queue(object):
    def __init__(self):
        self.__queue = LinkedList()

    def enqueue(self, item):
        """
        Method for adding a new item at the back of the queue

        :param item:
        :return:
        """
        self.__queue.append(item)

    def dequeue(self):
        """
        Method for returning an item from the front of the queue

        :return:
        """
        return self.__queue.pop_idx(0)

    def first(self):
        """
        Peek at the front of the queue without dequeueing it

        :return:
        """
        return self.__queue.front()

    def is_empty(self):
        """
        Boolean value on whether the queue is empty

        :return:
        """
        return self.__queue.is_empty()

    def __len__(self):
        """
        Return the length of the queue

        :return:
        """
        return len(self.__queue)

File: ch06/c/ll_queue.c
// This is a queue implementation backed by a linked list
// Relies on the linked_list functions
#include "ll_queue.h"

Queue *queue_new(void)
{
    // Constructs a new linked list and passes back the pointer
    Queue *queue = linkedlist_new();
    return queue;
}

void queue_enqueue(Queue *queue, int item)
{
    // Use append to append a new item
    linkedlist_append(queue, item);
}

int queue_dequeue(Queue *queue)
{
    // Use pop with idx = 0 to dequeue the head item
    int item = linkedlist_pop_idx(queue, 0);
    return item;
}

int queue_first(Queue *queue)
{
    // Use the peek function of the linkedlist to check the head
    int item = linkedlist_front(queue);
    return item;
}

bool queue_isempty(Queue *queue)
{
    // Use the is_empty function of our linkedlist
    return linkedlist_isempty(queue);
}

int queue_size(Queue *queue)
{
    // Use the size variable of our linkedlist
    return queue->size;
}

void queue_destroy(Queue *queue)
{
    // Use the linkedlist destructor
    linkedlist_destroy(queue);
}